File: AccuPay/Domain/Payroll/PhilHealthCalculator.vb
Option Strict On

Imports AccuPay.Entity
Imports PayrollSys

Namespace Global.AccuPay.Payroll

    Public Class PhilHealthCalculator

        Private ReadOnly _policy As PhilHealthPolicy
        Private ReadOnly _philHealthBrackets As ICollection(Of PhilHealthBracket)

        Private Const EcolaName As String = "ecola"

        Public Sub New(policy As PhilHealthPolicy, philHealthBrackets As ICollection(Of PhilHealthBracket))
            _policy = policy
            _philHealthBrackets = philHealthBrackets
        End Sub

        Public Sub Calculate(salary As Salary, paystub As Paystub, previousPaystub As Paystub, employee As Employee, payperiod As PayPeriod, allowances As ICollection(Of Allowance))
            ' Reset the PhilHealth to zero
            paystub.PhilHealthEmployeeShare = 0
            paystub.PhilHealthEmployerShare = 0

            Dim totalContribution As Decimal

            ' If auto compute the PhilHealth is true, then we use the available formulas to compute the total contribution.
            ' Otherwise, we use whatever amount is set in the salary.
            If salary.AutoComputePhilHealthContribution Then
                totalContribution = GetTotalContribution(
                    salary, paystub, previousPaystub, employee, allowances)
            Else
                totalContribution = salary.PhilHealthDeduction
            End If

            ' If totalContribution is zero, then the employee has no PhilHealth to pay
            If totalContribution <= 0 Then
                Return
            End If

            Dim halfContribution = AccuMath.Truncate(totalContribution / 2, 2)

            Dim employeeShare = halfContribution
            Dim employerShare = halfContribution

            ' Account for any division loss by putting the missing value to the employer's share
            If _policy.OddCentDifference Then
                Dim expectedTotal = halfContribution * 2
                Dim remainder As Decimal

                If expectedTotal < totalContribution Then
                    remainder = totalContribution - expectedTotal
                End If

                employerShare += remainder
            End If

            If employee.IsWeeklyPaid Then
                Dim is_deduct_sched_to_thisperiod = If(
                    employee.IsUnderAgency,
                    payperiod.PhHWeeklyAgentContribSched,
                    payperiod.PhHWeeklyContribSched)

                If is_deduct_sched_to_thisperiod Then
                    paystub.PhilHealthEmployeeShare = employeeShare
                    paystub.PhilHealthEmployerShare = employerShare
                End If
            Else
                Dim deductionSchedule = employee.PhilHealthSchedule

                If IsPhilHealthPaidOnFirstHalf(deductionSchedule, payperiod) Or IsPhilHealthPaidOnEndOfTheMonth(deductionSchedule, payperiod) Then
                    paystub.PhilHealthEmployeeShare = employeeShare
                    paystub.PhilHealthEmployerShare = employerShare
                ElseIf IsPhilHealthPaidPerPayPeriod(deductionSchedule) Then
                    paystub.PhilHealthEmployeeShare = employeeShare / CalendarConstants.SemiMonthlyPayPeriodsPerMonth
                    paystub.PhilHealthEmployerShare = employerShare / CalendarConstants.SemiMonthlyPayPeriodsPerMonth
                End If
            End If
        End Sub

        Private Function GetTotalContribution(salary As Salary,
                                              paystub As Paystub,
                                              previousPaystub As Paystub,
                                              employee As Employee,
                                              allowances As ICollection(Of Allowance)) As Decimal

            Dim calculationBasis = _policy.CalculationBasis

            Dim basisPay = 0D

            ' If philHealth calculation is based on the basic salary, get it from the salary record
            If calculationBasis = PhilHealthCalculationBasis.BasicSalary Then

                basisPay = PayrollTools.GetEmployeeMonthlyRate(employee, salary.BasicSalary)

            ElseIf calculationBasis = PhilHealthCalculationBasis.BasicAndEcola Then

                Dim monthlyRate = PayrollTools.GetEmployeeMonthlyRate(employee, salary.BasicSalary)

                Dim ecolas = allowances.
                    Where(Function(ea) ea.Product.PartNo.ToLower() = EcolaName)

                Dim ecolaPerMonth = 0D
                If ecolas.Any() Then
                    Dim ecola = ecolas.FirstOrDefault()

                    ecolaPerMonth = ecola.Amount * (employee.WorkDaysPerYear / CalendarConstants.MonthsInAYear)
                End If

                basisPay = monthlyRate + ecolaPerMonth

            ElseIf calculationBasis = PhilHealthCalculationBasis.Earnings Then

                basisPay = If(previousPaystub?.TotalEarnings, 0) + paystub.TotalEarnings

            ElseIf calculationBasis = PhilHealthCalculationBasis.GrossPay Then

                basisPay = If(previousPaystub?.GrossPay, 0) + paystub.GrossPay

            ElseIf calculationBasis = PhilHealthCalculationBasis.BasicMinusDeductions Then

                basisPay = If(previousPaystub?.TotalDaysPayWithOutOvertimeAndLeave, 0) + paystub.TotalDaysPayWithOutOvertimeAndLeave

            End If

            Dim totalContribution = ComputePhilHealth(basisPay)

            Return totalContribution
        End Function

        Private Function ComputePhilHealth(basis As Decimal) As Decimal
            Dim minimum = _policy.MinimumContribution
            Dim maximum = _policy.MaximumContribution
            Dim rate = _policy.Rate / 100

            ' Contribution should be bounded by the minimum and maximum
            Dim contribution = {{basis * rate, minimum}.Max(), maximum}.Min()
            ' Round to the nearest cent
            contribution = AccuMath.CommercialRound(contribution)

            Return contribution
        End Function

        <Obsolete>
        Private Function FindMatchingBracket(amount As Decimal) As PhilHealthBracket
            Return _philHealthBrackets.FirstOrDefault(
                Function(p) p.SalaryRangeFrom <= amount And p.SalaryRangeTo >= amount)
        End Function

        Private Function IsPhilHealthPaidOnFirstHalf(deductionSchedule As String, payperiod As PayPeriod) As Boolean
            Return payperiod.IsFirstHalf And (deductionSchedule = ContributionSchedule.FIRST_HALF)
        End Function

        Private Function IsPhilHealthPaidOnEndOfTheMonth(deductionSchedule As String, payperiod As PayPeriod) As Boolean
            Return payperiod.IsEndOfTheMonth And (deductionSchedule = ContributionSchedule.END_OF_THE_MONTH)
        End Function

        Private Function IsPhilHealthPaidPerPayPeriod(deductionSchedule As String) As Boolean
            Return deductionSchedule = ContributionSchedule.PER_PAY_PERIOD
        End Function

    End Class

End Namespace

File: Accupay.Data/Services/PayrollGeneration/PhilHealthCalculator.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Enums;
using AccuPay.Data.Helpers;
using AccuPay.Utilities;
using System;
using System.Collections.Generic;
using System.Linq;

namespace AccuPay.Data.Services
{
    public class PhilHealthCalculator
    {
        private readonly PhilHealthPolicy _policy;
        private readonly IReadOnlyCollection<PhilHealthBracket> _philHealthBrackets;

        public PhilHealthCalculator(PhilHealthPolicy policy, IReadOnlyCollection<PhilHealthBracket> philHealthBrackets)
        {
            _policy = policy;
            _philHealthBrackets = philHealthBrackets;
        }

        public void Calculate(Salary salary,
                            Paystub paystub,
                            Paystub previousPaystub,
                            Employee employee,
                            PayPeriod payperiod,
                            IReadOnlyCollection<Allowance> allowances,
                            string currentSystemOwner)
        {
            // Reset the PhilHealth to zero
            paystub.PhilHealthEmployeeShare = 0;
            paystub.PhilHealthEmployerShare = 0;

            decimal totalContribution;

            // If auto compute the PhilHealth is true, then we use the available formulas to compute the total contribution.
            // Otherwise, we use whatever amount is set in the salary.
            if (salary.AutoComputePhilHealthContribution)
                totalContribution = GetTotalContribution(salary,
                                                        paystub,
                                                        previousPaystub,
                                                        employee,
                                                        allowances,
                                                        currentSystemOwner);
            else
                totalContribution = salary.PhilHealthDeduction;

            // If totalContribution is zero, then the employee has no PhilHealth to pay
            if (totalContribution <= 0)
                return;

            var halfContribution = AccuMath.Truncate(totalContribution / 2, 2);

            var employeeShare = halfContribution;
            var employerShare = halfContribution;

            // Account for any division loss by putting the missing value to the employer's share
            if (_policy.OddCentDifference)
            {
                var expectedTotal = halfContribution * 2;
                decimal remainder = 0;

                if (expectedTotal < totalContribution)
                    remainder = totalContribution - expectedTotal;

                employerShare += remainder;
            }

            if (employee.IsWeeklyPaid)
            {
                var is_deduct_sched_to_thisperiod = employee.IsUnderAgency ? payperiod.PhHWeeklyAgentContribSched : payperiod.PhHWeeklyContribSched;

                if (is_deduct_sched_to_thisperiod)
                {
                    paystub.PhilHealthEmployeeShare = employeeShare;
                    paystub.PhilHealthEmployerShare = employerShare;
                }
            }
            else
            {
                var deductionSchedule = employee.PhilHealthSchedule;

                if (IsPhilHealthPaidOnFirstHalf(deductionSchedule, payperiod) ||
                    IsPhilHealthPaidOnEndOfTheMonth(deductionSchedule, payperiod))
                {
                    paystub.PhilHealthEmployeeShare = employeeShare;
                    paystub.PhilHealthEmployerShare = employerShare;
                }
                else if (IsPhilHealthPaidPerPayPeriod(deductionSchedule))
                {
                    paystub.PhilHealthEmployeeShare = employeeShare /
                                                CalendarConstants.SemiMonthlyPayPeriodsPerMonth;
                    paystub.PhilHealthEmployerShare = employerShare /
                                                CalendarConstants.SemiMonthlyPayPeriodsPerMonth;
                }
            }
        }

        private decimal GetTotalContribution(Salary salary,
                                            Paystub paystub,
                                            Paystub previousPaystub,
                                            Employee employee,
                                            IReadOnlyCollection<Allowance> allowances,
                                            string currentSystemOwner)
        {
            var calculationBasis = _policy.CalculationBasis;

            var basisPay = 0M;

            switch (calculationBasis)
            {
                case PhilHealthCalculationBasis.BasicSalary:
                    basisPay = PayrollTools.GetEmployeeMonthlyRate(employee, salary);
                    break;

                case PhilHealthCalculationBasis.BasicMinusDeductions:

                    if (employee.IsFixed)
                    {
                        basisPay = PayrollTools.GetEmployeeMonthlyRate(employee, salary);
                    }
                    else
                    {
                        basisPay = (previousPaystub?.TotalDaysPayWithOutOvertimeAndLeave ?? 0) +
                        paystub.TotalDaysPayWithOutOvertimeAndLeave;
                    }
                    break;

                case PhilHealthCalculationBasis.BasicMinusDeductionsWithoutPremium:

                    if (employee.IsFixed)
                    {
                        basisPay = PayrollTools.GetEmployeeMonthlyRate(employee, salary);
                    }
                    else
                    {
                        var totalHours = (previousPaystub?.TotalWorkedHoursWithoutOvertimeAndLeave ?? 0) +
                                    paystub.TotalWorkedHoursWithoutOvertimeAndLeave;

                        if (currentSystemOwner == SystemOwnerService.Benchmark && employee.IsPremiumInclusive)
                        {
                            totalHours = (previousPaystub?.RegularHoursAndTotalRestDay ?? 0) +
                                            paystub.RegularHoursAndTotalRestDay;
                        }

                        var monthlyRate = PayrollTools.GetEmployeeMonthlyRate(employee, salary);
                        var dailyRate = PayrollTools.GetDailyRate(monthlyRate, employee.WorkDaysPerYear);
                        var hourlyRate = PayrollTools.GetHourlyRateByDailyRate(dailyRate);

                        basisPay = totalHours * hourlyRate;
                    }
                    break;

                case PhilHealthCalculationBasis.BasicAndEcola:

                    var monthlyRate2 = PayrollTools.GetEmployeeMonthlyRate(employee, salary);

                    var ecolas = allowances.Where(ea => ea.Product.PartNo.ToLower() == ProductConstant.ECOLA);

                    var ecolaPerMonth = 0M;
                    if (ecolas.Any())
                    {
                        var ecola = ecolas.FirstOrDefault();

                        ecolaPerMonth = ecola.Amount * (employee.WorkDaysPerYear / CalendarConstants.MonthsInAYear);
                    }

                    basisPay = monthlyRate2 + ecolaPerMonth;
                    break;

                case PhilHealthCalculationBasis.Earnings:
                    basisPay = (previousPaystub?.TotalEarnings ?? 0) + paystub.TotalEarnings;
                    break;

                case PhilHealthCalculationBasis.GrossPay:
                    basisPay = (previousPaystub?.GrossPay ?? 0) + paystub.GrossPay;
                    break;

                default:
                    break;
            }

            var totalContribution = ComputePhilHealth(basisPay);

            return totalContribution;
        }

        private decimal ComputePhilHealth(decimal basis)
        {
            var minimum = _policy.MinimumContribution;
            var maximum = _policy.MaximumContribution;
            var rate = _policy.Rate / 100;

            // Contribution should be bounded by the minimum and maximum
            var contribution = new decimal[] {
                                new decimal[] { basis * rate, minimum }.Max(),
                                maximum
                            }.Min();
            // Round to the nearest cent
            contribution = AccuMath.CommercialRound(contribution);

            return contribution;
        }

        [Obsolete]
        private PhilHealthBracket FindMatchingBracket(decimal amount)
        {
            return _philHealthBrackets.FirstOrDefault(p => p.SalaryRangeFrom <= amount &&
                                                            p.SalaryRangeTo >= amount);
        }

        private bool IsPhilHealthPaidOnFirstHalf(string deductionSchedule, PayPeriod payperiod)
        {
            return payperiod.IsFirstHalf &&
                    deductionSchedule == ContributionSchedule.FIRST_HALF;
        }

        private bool IsPhilHealthPaidOnEndOfTheMonth(string deductionSchedule, PayPeriod payperiod)
        {
            return payperiod.IsEndOfTheMonth &&
                    deductionSchedule == ContributionSchedule.END_OF_THE_MONTH;
        }

        private bool IsPhilHealthPaidPerPayPeriod(string deductionSchedule)
        {
            return deductionSchedule == ContributionSchedule.PER_PAY_PERIOD;
        }
    }
}